File: sdk/javascript/src/resources/recommend.ts
import type { AxiosResponse } from "axios";

import type { RecoEngineClient } from "../client";
import type {
  BatchQuery,
  BatchRecommendResult,
  CacheStatus,
  FeedbackType,
  RecommendOptions,
  RecommendResult,
} from "../types";

function body(options: RecommendOptions) {
  return {
    top_k: options.topK ?? 10,
    filters: options.filters ?? {},
    include_raw_data: options.includeRawData ?? false,
  };
}

function withCache<T>(response: AxiosResponse<T>): T & { cache: CacheStatus | null } {
  const cache = (response.headers?.["x-cache"] as CacheStatus | undefined) ?? null;
  return { ...response.data, cache };
}

export class Recommend {
  constructor(private readonly client: RecoEngineClient) {}

  /** Items most similar to free text. */
  async byText(query: string, options: RecommendOptions = {}): Promise<RecommendResult> {
    return withCache(
      await this.client.request<Omit<RecommendResult, "cache">>({
        method: "POST",
        url: "/recommend/by-text",
        data: { query, ...body(options) },
      }),
    );
  }

  /** Items similar to one of your items; the item itself is never returned. */
  async byItem(externalId: string, options: RecommendOptions = {}): Promise<RecommendResult> {
    return withCache(
      await this.client.request<Omit<RecommendResult, "cache">>({
        method: "POST",
        url: "/recommend/by-item",
        data: { external_id: externalId, ...body(options) },
      }),
    );
  }

  /** Items matching a profile, e.g. `{ skills: "Python", experience: "5 years" }`. */
  async byProfile(profile: Record<string, string>, options: RecommendOptions = {}): Promise<RecommendResult> {
    return withCache(
      await this.client.request<Omit<RecommendResult, "cache">>({
        method: "POST",
        url: "/recommend/by-profile",
        data: { profile, ...body(options) },
      }),
    );
  }

  /** Up to 20 text queries in one request. */
  async batch(queries: BatchQuery[], options: Pick<RecommendOptions, "topK"> = {}): Promise<BatchRecommendResult> {
    return withCache(
      await this.client.request<Omit<BatchRecommendResult, "cache">>({
        method: "POST",
        url: "/recommend/batch",
        data: { queries, top_k: options.topK ?? 10 },
      }),
    );
  }

  /** Record a user's reaction to a result, using the `query_id` of the recommendation. */
  async submitFeedback(queryId: string, itemId: string, type: FeedbackType): Promise<void> {
    await this.client.request({
      method: "POST",
      url: "/recommend/feedback",
      data: { query_id: queryId, external_item_id: itemId, feedback_type: type },
    });
  }
}

File: sdk/python/recoengine/resources/recommend.py
"""Recommendations and feedback."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Dict, List, Optional, Sequence, Union

import httpx

from recoengine._base import recommend_body
from recoengine.models import BatchQuery, BatchRecommendResult, FeedbackType, RecommendResult

if TYPE_CHECKING:
    from recoengine.client import AsyncRecoEngineClient, RecoEngineClient

Filters = Optional[Dict[str, Any]]
BatchQueries = Sequence[Union[BatchQuery, Dict[str, Any]]]


def _result(response: httpx.Response) -> RecommendResult:
    return RecommendResult.model_validate(
        {**response.json(), "cache": response.headers.get("x-cache")}
    )


def _batch_body(queries: BatchQueries, top_k: int) -> dict[str, Any]:
    items: List[Dict[str, Any]] = [
        q.model_dump() if isinstance(q, BatchQuery) else dict(q) for q in queries
    ]
    return {"queries": items, "top_k": top_k}


def _feedback_body(query_id: str, item_id: str, feedback_type: FeedbackType) -> dict[str, Any]:
    return {"query_id": query_id, "external_item_id": item_id, "feedback_type": feedback_type}


class Recommend:
    def __init__(self, client: RecoEngineClient) -> None:
        self._client = client

    def by_text(
        self,
        query: str,
        *,
        top_k: int = 10,
        filters: Filters = None,
        include_raw_data: bool = False,
    ) -> RecommendResult:
        """Items most similar to free text.

        Filters use your `filter_fields`: `"Delhi"` exact, `["Delhi", "Pune"]` any of,
        `{"gte": 3, "lte": 8}` range.
        """
        body = {"query": query, **recommend_body(top_k, filters, include_raw_data)}
        return _result(self._client._request("POST", "/recommend/by-text", json=body))

    def by_item(
        self,
        external_id: str,
        *,
        top_k: int = 10,
        filters: Filters = None,
        include_raw_data: bool = False,
    ) -> RecommendResult:
        """Items similar to one of your items; the item itself is never returned."""
        body = {"external_id": external_id, **recommend_body(top_k, filters, include_raw_data)}
        return _result(self._client._request("POST", "/recommend/by-item", json=body))

    def by_profile(
        self,
        profile: Dict[str, str],
        *,
        top_k: int = 10,
        filters: Filters = None,
        include_raw_data: bool = False,
    ) -> RecommendResult:
        """Items matching a profile, e.g. {"skills": "Python", "experience": "5 years"}."""
        body = {"profile": profile, **recommend_body(top_k, filters, include_raw_data)}
        return _result(self._client._request("POST", "/recommend/by-profile", json=body))

    def batch(self, queries: BatchQueries, *, top_k: int = 10) -> BatchRecommendResult:
        """Up to 20 text queries in one request."""
        response = self._client._request(
            "POST", "/recommend/batch", json=_batch_body(queries, top_k)
        )
        return BatchRecommendResult.model_validate(
            {**response.json(), "cache": response.headers.get("x-cache")}
        )

    def submit_feedback(self, query_id: str, item_id: str, feedback_type: FeedbackType) -> None:
        """Record a user's reaction to a result, using the result's `query_id`."""
        self._client._request(
            "POST", "/recommend/feedback", json=_feedback_body(query_id, item_id, feedback_type)
        )


class AsyncRecommend:
    def __init__(self, client: AsyncRecoEngineClient) -> None:
        self._client = client

    async def by_text(
        self,
        query: str,
        *,
        top_k: int = 10,
        filters: Filters = None,
        include_raw_data: bool = False,
    ) -> RecommendResult:
        body = {"query": query, **recommend_body(top_k, filters, include_raw_data)}
        return _result(await self._client._request("POST", "/recommend/by-text", json=body))

    async def by_item(
        self,
        external_id: str,
        *,
        top_k: int = 10,
        filters: Filters = None,
        include_raw_data: bool = False,
    ) -> RecommendResult:
        body = {"external_id": external_id, **recommend_body(top_k, filters, include_raw_data)}
        return _result(await self._client._request("POST", "/recommend/by-item", json=body))

    async def by_profile(
        self,
        profile: Dict[str, str],
        *,
        top_k: int = 10,
        filters: Filters = None,
        include_raw_data: bool = False,
    ) -> RecommendResult:
        body = {"profile": profile, **recommend_body(top_k, filters, include_raw_data)}
        return _result(await self._client._request("POST", "/recommend/by-profile", json=body))

    async def batch(self, queries: BatchQueries, *, top_k: int = 10) -> BatchRecommendResult:
        response = await self._client._request(
            "POST", "/recommend/batch", json=_batch_body(queries, top_k)
        )
        return BatchRecommendResult.model_validate(
            {**response.json(), "cache": response.headers.get("x-cache")}
        )

    async def submit_feedback(
        self, query_id: str, item_id: str, feedback_type: FeedbackType
    ) -> None:
        await self._client._request(
            "POST", "/recommend/feedback", json=_feedback_body(query_id, item_id, feedback_type)
        )
